### PR TITLE
ci: speed up PR test workflow

### PR DESCRIPTION
**Pull request by zereight**
## Summary
- Remove the duplicate `coverage` job that re-ran `test:mock` without collecting coverage
- Drop redundant `test:remote-auth` CI step (already included in `test:mock`)
- Slim `integration-test` to Docker build only and run it in parallel with `test` (removed `needs: test` and duplicate oauth/list-mr steps)
- Parallelize `test:mock` via `scripts/run-mock-tests.sh` using `node --test --test-concurrency=4`

## Test plan
- [x] `npm run test:mock` passes locally (~80s vs ~4m before)
- [ ] PR checks: `test`, `integration-test`, `code-quality`, `docs-build` all green
- [ ] Confirm total PR workflow wall time drops from ~6m to ~2-3m

Made with [Cursor](https://cursor.com)

### CODE REVIEW

File: scripts/run-mock-tests.sh [hunk 1]
```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+# ponytail: one node --test invocation with concurrency beats 40+ sequential shells
+node --import tsx/esm --test --test-concurrency=4 \
+  test/path-segment-encoding.test.ts \
```
[the remark this answers is not in the document]
Comment (zereight) [reply scripts/run-mock-tests.sh]: Fixed in 9e4f6d1 and 3a67bd3.

- `findMockServerPort()` now returns `0` (placeholder only)
- `MockGitLabServer.start()` binds with `listen(0)` and reads the OS-assigned port from `server.address()`
- Parallel `node --test --test-concurrency=4` no longer races on pre-reserved ports

Verified locally with `npm run test:mock`.

File: scripts/run-mock-tests.sh [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  test/utils/proxy-client-ip.test.ts \
+  test/utils/forwarded-public-base-url.test.ts \
+  test/utils/graphql-query.test.ts \
+  test/utils/wiki-title.test.ts \
+  test/utils/merge-request-position.test.ts \
+  test/utils/redact-sensitive.test.ts \
+  test/utils/version-check.test.ts \
+  test/nullish-tool-arguments-schema.test.ts \
+  test/test-ci-variables.ts \
+  test/test-dependency-proxy.ts
```
[the remark this answers is not in the document]
Comment (zereight) [reply scripts/run-mock-tests.sh]: Addressed in 4e682cb.

Replaced the hard-coded file list with `find`-based discovery in `scripts/run-mock-tests.sh`, excluding live/schema/manual suites only. New mock tests under `test/` are picked up automatically.

File: test/utils/mock-gitlab-server.ts [hunk 3]
```diff
@@ -649,46 +649,31 @@ export class MockGitLabServer {
 }
 
 /**
- * Helper to find available port for mock server
+ * Helper to find available port for mock server (OS-assigned ephemeral port).
  */
-export async function findMockServerPort(
-  basePort: number = 9000,
-  maxAttempts: number = 10
-): Promise<number> {
```
[the remark this answers is not in the document]
Comment (zereight) [reply test/utils/mock-gitlab-server.ts:657]: Fixed in 3a67bd3.

Removed the bind-then-close port probe. `MockGitLabServer.start()` now listens on port `0` and updates `config.port` from `server.address()` after bind, so there is no window where another parallel test can steal the port.

Also updated `test-ci-variables.ts` and `test-dependency-proxy.ts` to read the API URL from `mockServer.getUrl()` after `start()`.

[the remark this answers is not in the document]
Comment (zereight) [reply test/utils/mock-gitlab-server.ts:657]: CI was still failing after the port-0 bind fix: GitHub Actions hit intermittent `Unable to deserialize cloned data` errors when multiple server-spawning suites ran under `xargs -P 4`.

Fixed in 7ab6ccc by splitting the runner:
- unit tests (`test/utils`, stateless helpers, etc.) stay parallel (`-P 4`)
- MCP/mock server suites run sequentially (`-P 1`, `--test-concurrency=1`)

Local `npm run test:mock` passes with this layout.

File: scripts/run-mock-tests.sh [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  ! -name 'mock-gitlab-server.ts' ! -name 'server-launcher.ts' \
+  ! -name 'oauth-tests.ts' ! -name 'schema-tests.ts' ! -name 'test-json-schema.ts' \
+  ! -name 'dynamic-api-url-test.ts' ! -name 'dynamic-routing-tests.ts' \
+  ! -name 'callback-proxy-tests.ts' ! -name 'test-all-transport-server.ts' \
+  ! -name 'multi-server-test.ts' ! -name 'client-pool-test.ts' \
+  ! -name 'no-proxy-test.ts' ! -name 'no-proxy-integration-test.ts' \
+  ! -name 'remote-auth-tests.ts' ! -name 'test-mr-diffs-filter.ts' \
+  ! -name 'test-mr-file-diffs.ts' ! -name 'test-token-optimizations.ts' \
+  ! -name 'test-merge-request-approvals.ts' ! -name 'config-allowed-groups.test.ts' \
+  -print0 | sort -z | xargs -0 node --import tsx/esm --test --test-concurrency=4
```
[the remark this answers is not in the document]
Comment (zereight) [reply scripts/run-mock-tests.sh]: Acknowledged. These suites were already excluded from `test:mock` before this PR (same `EXCLUDE_OPTS` at merge-base). They are manual/live or covered elsewhere (`test:schema`, `live-tests` workflow). This PR only changes how discovered mock suites run (parallel unit + sequential server tests) and does not widen/narrow the discovery set. Keeping exclusions as-is to avoid inflating PR CI time; we can add a follow-up if we want those suites back in `test:mock`.
